<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Usage · BioDemuX.jl</title><meta name="title" content="Usage · BioDemuX.jl"/><meta property="og:title" content="Usage · BioDemuX.jl"/><meta property="twitter:title" content="Usage · BioDemuX.jl"/><meta name="description" content="Documentation for BioDemuX.jl."/><meta property="og:description" content="Documentation for BioDemuX.jl."/><meta property="twitter:description" content="Documentation for BioDemuX.jl."/><meta property="og:url" content="https://I-Mihara.github.io/BioDemuX.jl/usage/"/><meta property="twitter:url" content="https://I-Mihara.github.io/BioDemuX.jl/usage/"/><link rel="canonical" href="https://I-Mihara.github.io/BioDemuX.jl/usage/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioDemuX.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Usage</a><ul class="internal"><li><a class="tocitem" href="#Dependencies"><span>Dependencies</span></a></li><li><a class="tocitem" href="#Basic-Usage"><span>Basic Usage</span></a></li></ul></li><li><a class="tocitem" href="../cli/">CLI</a></li><li><a class="tocitem" href="../options/">Options</a></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Usage</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Usage</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/I-Mihara/BioDemuX.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/I-Mihara/BioDemuX.jl/blob/main/docs/src/usage.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Usage"><a class="docs-heading-anchor" href="#Usage">Usage</a><a id="Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Usage" title="Permalink"></a></h1><h2 id="Dependencies"><a class="docs-heading-anchor" href="#Dependencies">Dependencies</a><a id="Dependencies-1"></a><a class="docs-heading-anchor-permalink" href="#Dependencies" title="Permalink"></a></h2><ul><li><strong>Julia</strong> &gt;= 1.10 (includes the <code>Distributed</code> standard library)</li><li><strong>Julia packages</strong><ul><li>DataFrames &gt;= 1.7.0</li><li>CSV &gt;= 0.10.14</li><li>CodecZlib &gt;= 0.7.0</li><li>BufferedStreams &gt;= 1.2.2</li></ul></li></ul><h2 id="Basic-Usage"><a class="docs-heading-anchor" href="#Basic-Usage">Basic Usage</a><a id="Basic-Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-Usage" title="Permalink"></a></h2><p>The primary function of this package is <code>execute_demultiplexing()</code>. It classifies sequences in a FASTQ file by aligning them with reference barcodes in barcode file. Usage is as follows:</p><pre><code class="language-Julia hljs">using BioDemuX
execute_demultiplexing(FASTQ_file, barcode_file, output_directory)</code></pre><h3 id="Input"><a class="docs-heading-anchor" href="#Input">Input</a><a id="Input-1"></a><a class="docs-heading-anchor-permalink" href="#Input" title="Permalink"></a></h3><h4 id="FASTQ-File"><a class="docs-heading-anchor" href="#FASTQ-File">FASTQ File</a><a id="FASTQ-File-1"></a><a class="docs-heading-anchor-permalink" href="#FASTQ-File" title="Permalink"></a></h4><ul><li>There is no restriction on the sequence length in the FASTQ file.</li><li>The input can be gzipped; the function automatically detects and processes the files accordingly.</li><li>The function can take one or two FASTQ files as input. In the case of using two FASTQ files, the command can be executed as follows:</li></ul><pre><code class="language-julia hljs">execute_demultiplexing(FASTQ_file1, FASTQ_file2, barcode_file, output_directory)</code></pre><p>When using two FASTQ files, sequences in the <code>FASTQ_file2</code> are classified based on the alignment of the <code>FASTQ_file1</code> sequences with the barcodes in the barcode reference file. Hence, the corresponding reads in both FASTQ files must be in the same order and present in equal numbers.</p><h4 id="Barcode-Reference-File"><a class="docs-heading-anchor" href="#Barcode-Reference-File">Barcode Reference File</a><a id="Barcode-Reference-File-1"></a><a class="docs-heading-anchor-permalink" href="#Barcode-Reference-File" title="Permalink"></a></h4><ul><li>The reference file is expected to be a CSV or TSV file containing the following columns: <code>ID</code>, <code>Full_seq</code>, <code>Full_annotation</code>, as shown below:</li></ul><pre><code class="nohighlight hljs">ID  Full_seq	Full_annotation
001-barcode ACAGACUACAAA XXXBBBBBBBXX</code></pre><ul><li><p>In the <code>Full_seq</code> column, the region specified as <code>B</code> in the <code>Full_annotation</code> column is considered as the barcode.</p></li><li><p>Alternatively, a FASTA file of barcode sequences can be used as the reference. In this case, each sequence in the FASTA file is treated as a full barcode (the entire sequence is considered the barcode region) and the header line of each entry (without the <code>&gt;</code> prefix) is used as its <code>ID</code>.</p></li></ul><h3 id="Output"><a class="docs-heading-anchor" href="#Output">Output</a><a id="Output-1"></a><a class="docs-heading-anchor-permalink" href="#Output" title="Permalink"></a></h3><ul><li>All output files will be saved in the specified <code>output_directory</code>.</li><li>The output is gzipped depending on the input FASTQ format, or can be specified using the <code>gzip_output</code> option.</li><li>The names of the output files are based on the filename of the FASTQ file as the prefix and the <code>ID</code> values in the barcode reference file. For example, if the FASTQ filename is <code>sample.fastq</code> and the reference file contains IDs such as <code>001</code> and <code>002</code>, the resulting output files will be named <code>sample.001.fastq</code>, <code>sample.002.fastq</code>, and so on. You can freely change the prefix by specifying the <code>output_prefix</code> argument.</li><li>Sequences that do not match any barcode in the reference file are saved in <code>unknown.fastq</code>. Sequences that have ambiguous classification (i.e., they match multiple barcodes with similar scores) are saved in <code>ambiguous_classification.fastq</code>. These FASTQ files also have prefix like <code>sample.unknown.fastq</code> and <code>sample.ambiguous_classification.fastq</code>.</li><li>If the <code>output_directory</code> does not exist, a new directory is created to store the output files.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../cli/">CLI »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Thursday 18 December 2025 22:14">Thursday 18 December 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
